### ISSUE TITLE
Update multiple list items in batch

### ISSUE THREAD

**Issue by Saissaken** [SharePoint/PnP-JS-Core#356]
#### Category
[ ] Enhancement

[ ] Bug

[x ] Question

#### Expected / Desired Behavior / Question
How can I update multiple items at once?

My code is the following:

```js
import pnp from 'sp-pnp-js';

let list = pnp.sp.web.lists.getByTitle("Sample List");

list.getListItemEntityTypeFullName().then(entityTypeFullName => {
    let batch = pnp.sp.web.createBatch();
    list.items.getById(1).inBatch(batch).update({ Title: "Batch update 1" }, entityTypeFullName).then(b => {
        console.log("Finish first");
    });

    list.items.getById(2).inBatch(batch).update({ Title: "Batch update 2" }, entityTypeFullName).then(b => {
        console.log("Finish second");
    });

    batch.execute().then(d => console.log("Done"));
});
```

When I executed this code, the browser did the same GET request 3 times to this endpoint: 

_api/web/lists/getByTitle('Sample%20List')?$select=ListItemEntityTypeFullName

 to obtain the list entity, but there are no update requests.

I already batched .get() and .add() requests and it works great, I'm only having problems with .update().

Thanks!

**Edit:** I'm using sp-pnp-js 2.0.1, Sharepoint Online and Chrome

**Comment by patrick-rodgers** [comment]
I think the issue is that update doesn't take the entity type name as a second parameter, it takes an etag. So you are trying to update based on the etag. The signature of that method is:

```TypeScript
public update(properties: TypedHash<any>, eTag = "*"): Promise<ItemUpdateResult> {
```

Try leaving off the second parameter and see if things work for you. We don't currently take in the entity type name for updates.

**Comment by Saissaken** [comment]
Makes sense, but I tried removing the second parameter and it didn't work. I also tried with a fresh list, and tried building the url like this:

> new Web(appWebUrl);

I updated my code to this:

```js
import {Web} from 'sp-pnp-js';
import pnp from 'sp-pnp-js';

let appWebUrl = pnp.util.getUrlParamByName("SPAppWebUrl");
let web = new Web(appWebUrl);
let list = web.lists.getByTitle("Test List");
```

To remove any doubts, yes, the list "Test List" exists, and the IDs of the items are correct.

And, using that code, this are the things I tried:

**Get**
```js
let batch = web.createBatch();

list.items.getById(1).inBatch(batch).get().then(d => console.log(d));
list.items.getById(2).inBatch(batch).get().then(d => console.log(d));
list.items.getById(3).inBatch(batch).get().then(d => console.log(d));

batch.execute().then(d => console.log("Done"));
```
Results: Worked perfectly

---

**Add** 
```js
let batch = web.createBatch();

list.items.inBatch(batch).add({Title: "Batch add 1"}).then(d => console.log(d));
list.items.inBatch(batch).add({Title: "Batch add 2"}).then(d => console.log(d));
list.items.inBatch(batch).add({Title: "Batch add 3"}).then(d => console.log(d));

batch.execute().then(d => console.log("Done"));
```
Results: Fails.

1. Requested the url "/_api/web/lists/getByTitle('Test%20List')?$select=ListItemEntityTypeFullName" 3 times
2. Requested the "/_api/$batch" with this data (I replaced the urls with ***, but they are correct):
```
--batch_c17f55d4-3bb8-46ac-b2e4-0b165e73d382
Content-Type: multipart/mixed; boundary="changeset_630d2e8e-8666-48c2-a6c5-73cd6ed373fe"

--changeset_630d2e8e-8666-48c2-a6c5-73cd6ed373fe
Content-Type: application/http
Content-Transfer-Encoding: binary

POST https://***/_api/web/lists/getByTitle('Test List')/items HTTP/1.1
Accept: application/json;
Content-Type: application/json;odata=verbose;charset=utf-8

{"__metadata":{"type":"SP.Data.Test_x0020_ListListItem"},"Title":"Batch add 1"}

--changeset_630d2e8e-8666-48c2-a6c5-73cd6ed373fe
Content-Type: application/http
Content-Transfer-Encoding: binary

POST  https://***/_api/web/lists/getByTitle('Test List')/items HTTP/1.1
Accept: application/json;
Content-Type: application/json;odata=verbose;charset=utf-8

{"__metadata":{"type":"SP.Data.Test_x0020_ListListItem"},"Title":"Batch add 2"}

--changeset_630d2e8e-8666-48c2-a6c5-73cd6ed373fe--

--batch_c17f55d4-3bb8-46ac-b2e4-0b165e73d382--
```
Note that **the batch request is missing the third changeset.** So the first and second item gets created but the third one no. And throws this error: 

> Uncaught (in promise) BatchParseException: Could not properly parse responses to match requests in batch.

So I tried again but with this code: 

```js
list.getListItemEntityTypeFullName().then(entityTypeFullName => {
    let batch = web.createBatch();

    list.items.inBatch(batch).add({Title: "Batch add 1"}, entityTypeFullName).then(d => console.log(d));
    list.items.inBatch(batch).add({Title: "Batch add 2"}, entityTypeFullName).then(d => console.log(d));
    list.items.inBatch(batch).add({Title: "Batch add 3"}, entityTypeFullName).then(d => console.log(d));

    batch.execute().then(d => console.log("Done"));
});
```
Results: It worked perfectly

---

Now back to the **.update()**
```js
let batch = web.createBatch();

list.items.getById(1).inBatch(batch).update({Title: "Updated from batch 1"}).then(d => console.log(d));
list.items.getById(2).inBatch(batch).update({Title: "Updated from batch 2"}).then(d => console.log(d));
list.items.getById(3).inBatch(batch).update({Title: "Updated from batch 3"}).then(d => console.log(d));

batch.execute().then(d => console.log("Done"));
```
Results: Fails. It load "/_api/web/lists/getByTitle('Test%20List')?$select=ListItemEntityTypeFullName" 3 times, and didn't do nothing more. Not event the "/_api/$batch" call.

---

Now an **.update()** without batch:

```js
let batch = web.createBatch();

list.items.getById(1).inBatch(batch).update({Title: "Updated from batch 1"}).then(d => console.log(d));
list.items.getById(2).inBatch(batch).update({Title: "Updated from batch 2"}).then(d => console.log(d));
list.items.getById(3).inBatch(batch).update({Title: "Updated from batch 3"}).then(d => console.log(d));

batch.execute().then(d => console.log("Done"));
```
Results: Works perfectly

Am I doing something wrong?

**Comment by patrick-rodgers** [comment]
Shoot - no you aren't doing anything wrong. I caught this bug myself in add but it is also appearing in update. The issue is a race condition (which you can also call bad coding on my part :)) where the batch dependency is resolved and the final item isn't added to the batch before it executes. I'll have a look and get this sorted for tomorrow's release.

**Comment by patrick-rodgers** [comment]
Fixed in #357.
